LFSC LRAT Output

alex-ozdemir opened the pull request:
In this PR I implement LFSC output for LRAT proofs.

I also added a unit test for this output.

The natural reviewers seemed like Yoni (since he knows LFSC) and Andres (since he's reviewed the rest of the `LratProof` class). Since I just dropped a painfully huge diff on Andres, I've assigned Yoni, but Yoni, if you want to call in someone else to be another pair of eyes, feel free.

4tXJ7f commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,112 @@
+    {
+      lfscWithoutWhitespace << c;
+    }
+  }
```
Suggestion: Create a helper function for this.

alex-ozdemir replied:
Good idea! This is also an opportunity to use some STL.

4tXJ7f replied:
Nice :)

4tXJ7f commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,112 @@
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Black box testing of the DRAT proof class
```
```suggestion
 ** \brief Black box testing of the LRAT proof class
```

4tXJ7f commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,112 @@
+ **
+ ** \brief Black box testing of the DRAT proof class
+ **
+ ** In particular, tests DRAT binary parsing.
```
```suggestion
 ** In particular, tests LRAT LFSC output.
```

4tXJ7f commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,112 @@
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
```
```suggestion
 ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
```

4tXJ7f commented on src/proof/lrat/lrat_proof.cpp:
```diff
@@ -75,6 +75,54 @@ std::ostream& operator<<(std::ostream& o, const LratUPTrace& trace)
+/**
+ * Print a list of clause indices to go to while doing UP.
+ *
+ * i.e. a value of type Trace
```
Add comments for the `@param`s.

alex-ozdemir commented:
Thanks for the feedback, Andres! I believe that I've addressed all your comments. If that's not the case, or if you think further work should be done to address any of them, let me know.

yoni206 commented on src/proof/lrat/lrat_proof.cpp:
```diff
@@ -75,6 +75,63 @@ std::ostream& operator<<(std::ostream& o, const LratUPTrace& trace)
+ */
+void printIndices(std::ostream& o, const std::vector<ClauseIdx>& indices)
+{
+  for (ClauseIdx idx : indices)
```
Perhaps add an assertion that the list is sorted (as assumed by `lrat.plf`)?

yoni206 commented on src/proof/lrat/lrat_proof.h:
```diff
@@ -52,6 +52,15 @@ class LratInstruction
    * @param out the stream to write to
    */
   virtual void outputAsText(std::ostream& out) const = 0;
```
What tool can be used to process `out` after the operation of `outputAsText`, if any? Does it correspond to the format of the examples in the LRAT paper?

alex-ozdemir replied:
Good question. The answer is the `ClausalBitVectorProof` (in review) and the `ProofManager`.

Since this method just writes an LFSC LRAT proof **value** (not a proof!), the `ClausalBitVectorProof` will write the axiom which uses a side condition to check the proof value and then prove bottom.

The `ProofManager` also needs to write a bunch of closing parentheses after that.

Unfortunately the answer is yes -- the proof manager (or some other component) write a bunch of closing parentheses to the stream.

yoni206 commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,113 @@
+  return out;
+}
+
+void LfscProofBlack::testOutputThreeAsLfsc()
```
What does `Three` stand for in `testOutputThreeAsLfsc`?

alex-ozdemir replied:
Ahh -- three lines. I'll remove that detail. It made sense in a different context, but not here.

yoni206 commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,113 @@
+{
+  std::vector<std::unique_ptr<LratInstruction>> instructions;
+
+  // 6   d 1 2
```
Is this proof taken from some paper or something? If so, it might be useful to include lines 1--5 (which I assume are the original clauses?).

alex-ozdemir replied:
Ah -- I actually just make this proof up.

I didn't think it was important to have a semantically valid proof, since this test is really just about formatting.

yoni206 commented on test/unit/proof/lrat_proof_black.h:
```diff
@@ -0,0 +1,113 @@
+      "    RATHintsn)) "
+      "LRATProofn)))";
+
+  TS_ASSERT_EQUALS(filterWhitespace(lfsc.str()), filterWhitespace(expectedLfsc));
```
What do you think about applying `filterWhitespace` on both strings, in case in the future someone adds a space somewhere?

alex-ozdemir replied:
I think that's a great idea!

alex-ozdemir commented:
Thanks for the review Yoni!

I think that I've addressed or responded to each of your comments. Let me know if I missed any, or if you think any could be better addressed.
